alf/environments/wrappers.py

```diff
@@ -12,187 +12,166 @@
+            observation_space = traverse_transform_space(
+                space=observation_space,
+                levels=levels,
+                field=field,
+                func=self.transform_space)
+        self.observation_space = observation_space
+
+    def observation(self, observation):
+        for levels, field in self._paths:
+            observation = common.traverse_transform_observation(
+                obs=observation,
+                levels=levels,
+                field=field,
+                func=self.transform_observation)
+        return observation
+
+    def transform_space(self, observation_space, field=None):
+        """Transform space
+
+        Subclass should override this to perform transformation
+
+        Args:
+             observation_space (gym.Space): space to be transformed
+             field (str): field to be transformed
+        """
+        return observation_space
+
+    def transform_observation(self, observation, field=None):
+        """Transform observation
+
+        Subclass should override this to perform transformation
+
+        Args:
+             observation (ndarray): observation to be transformed
+             field (str): field to be transformed
+        """
+        return observation
+
+
+@gin.configurable
+class FrameStack(BaseObservationWrapper):
     """Stack previous `stack_size` frames, applied to Gym env."""
 
-    def __init__(
-            self,
-            env,
-            stack_size=4,
-            channel_order='channels_last',
-            fields_to_stack=None,
-    ):
+    def __init__(self,
+                 env,
+                 stack_size=4,
+                 channel_order='channels_last',
+                 fields=None):
         """Create a FrameStack object.
 
         Args:
             env (gym.Space): gym environment.
             stack_size (int):
-            channel_order (str): one of `channels_last` or `channels_first`.
-                The ordering of the dimensions in the images.
-                `channels_last` corresponds to images with shape
-                `(height, width, channels)` (Atari's default) while
-                `channels_first` corresponds to images with shape
-                `(channels, height, width)`.
-            fields_to_stack (list of str): optional paths to the fields of the
+            channel_order (str): The ordering of the dimensions in the images.
+                should be one of `channels_last` or `channels_first`.
+            fields (list[str]): optional paths to the fields of the
                 Dict env.observation_space.  If specified, only use the
-                spaces corresponding to the keys in FrameStack.  If is None,
-                stack all recognized spaces.
-                When input space is not a Dict, we just try to stack if we can.
-                Fields_to_stack doesn't apply anymore.
+                spaces corresponding to the keys in FrameStack.
```

Comment:
If fields is None, then no field is transformed.  This is different from the current assumption of FrameStack (all fields are stacked, if fields is None).  Please add a comment about what happens if fields is None.  Also please double check to make sure current gin files have the right fields specified.  I think ac_target_nav is the only one, but I may be wrong.

Reply:
Agree. Need to comment the behavior of "fields=None", which is to transform the non-dict observation. In this case observation cannot be dictionary.
In general, the comment of a function needs to contain enough information for the user to use the function correctly.

Reply:
comment updated